### PR TITLE
rustdoc: list matching impls on type aliases

### LINKED ISSUE #32077: Rustdoc doesn't handle impl section of type definitions
In [rust-xcb](https://github.com/rtbo/rust-xcb) I have the following code:

``` rust
/// a key was pressed/released
pub type KeyPressEvent = base::Event<xcb_key_press_event_t>;

impl KeyPressEvent {
    /// The keycode (a number representing a physical key on the keyboard) of the key
    /// which was pressed.
    pub fn detail(&self) -> Keycode { ... }

    /// Time when the event was generated (in milliseconds).
    pub fn time(&self) -> Timestamp { ... }

    ...
}
```

However as one can see [here](http://rtbo.github.io/rust-xcb/xcb/xproto/type.KeyPressEvent.html), the impl section is ignored.

### LINKED ISSUE #99952: [rustdoc] Generate documentation for type definitions
Would it be possible to add traits and functions to type definitions? For example, it's really hard to figure out what functions exist on this type: https://docs.rs/blake2/0.10.4/blake2/type.Blake2b.html
It's a type definition of a generic struct where some of the impls on the generic struct depend on the traits the generic implements. 
so having docs on the type definition will allow seeing what traits/functions end up implemented on that type

### PR DESCRIPTION
**Pull request by notriddle**
Fixes https://github.com/rust-lang/rust/issues/32077

Fixes #99952

Remake of https://github.com/rust-lang/rust/pull/112429

Partially reverts https://github.com/rust-lang/rust/pull/112543, but keeps the test case.

This version of the PR avoids the infinite loop by structurally matching types instead of using full unification. This version does not support type alias trait bounds, but the compiler does not enforce those anyway (https://github.com/rust-lang/rust/issues/21903).

r? @GuillaumeGomez 

CC @lcnr

### CODE REVIEW

**Review by GuillaumeGomez — APPROVED** [review]

File: src/librustdoc/html/render/mod.rs [hunk 1]
```diff
@@ -1139,8 +1141,41 @@ fn render_assoc_items_inner(
...
+        let Some(mut av) = cache.impls.get(&aliased_type_defid).cloned() &&
+        let Some(alias_def_id) = containing_item.item_id.as_def_id()
+    {
+        // This branch of the compiler compares types structually, but does
+        // not check trait bounds. That's probably fine, since type aliases
+        // don't normally constrain on them anyway.
+        // https://github.com/rust-lang/rust/issues/21903
+        //
+        // FIXME(lazy_type_alias): Once the feature is complete or stable, rewrite this to use type unification.
+        // Be aware of `tests/rustdoc/issue-32077-type-alias-impls.rs` which might regress.
```
Comment (fmease) [inline src/librustdoc/html/render/mod.rs:1157]:
```suggestion
        // Be aware of `tests/rustdoc/issue-112515-impl-ty-alias.rs` which might regress.
```

Sorry, copied the wrong path from the PR :sweat_smile:

File: src/librustdoc/html/render/mod.rs [hunk 2]
```diff
@@ -1139,8 +1141,41 @@ fn render_assoc_items_inner(
...
+        let Some(alias_def_id) = containing_item.item_id.as_def_id()
+    {
+        // This branch of the compiler compares types structually, but does
+        // not check trait bounds. That's probably fine, since type aliases
+        // don't normally constrain on them anyway.
+        // https://github.com/rust-lang/rust/issues/21903
+        //
+        // FIXME(lazy_type_alias): Once the feature is complete or stable, rewrite this to use type unification.
+        // Be aware of `tests/rustdoc/issue-112515-impl-ty-alias.rs` which might regress.
+        // unification.
```
Comment (lcnr) [inline src/librustdoc/html/render/mod.rs:1158]:
```suggestion
```

**Review by lcnr — COMMENTED** [review]
from a types team pov this is fine, don't know the code structure of rustdoc well enough to review the rest of this change

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Time to merge it then. Thanks everyone!

@bors r+

**Comment by steffahn** [comment]
Probably intended (and it probably *is* more useful this way), but “fun” side-effect (of this general change, I suppose): The documentation of `nalgebra` (default features, built locally from the same source) casually grows from 27MB to 515MB between `1.72` and `nightly 62ebe3a2b 2023-09-08`. Admitted, they are combining [a ton of type aliases](https://docs.rs/nalgebra/latest/nalgebra/base/index.html#types) with [a ton of `impl`s](https://docs.rs/nalgebra/latest/nalgebra/base/struct.Matrix.html), so it’s all nice and quadratic.

Also, is it intended that none of the `impl`s and methods appear in the side-bar? This makes navigating, especially pages with lots of contents, rather hard.

**Comment by steffahn** [comment]
Though this makes me wonder whether perhaps an opt-out option is wanted. For `ndalgebra`, they might want to opt-out of increasing the docs size at least for the deprecated aliases they have, which on first glance feels like it’s almost half of their type aliases (i.e. perhaps truly about half of the HTML content?)

For `typenum` which has a similar magnitude of size increase (44MB to 523MB for me), it seems like the additional `impl`s on [***all*** the constants](https://docs.rs/typenum/latest/typenum/consts/index.html) it probably not worth the additional file size (and documentation time either).

**Comment by GuillaumeGomez** [comment]
> Also, is it intended that none of the impls and methods appear in the side-bar? This makes navigating, especially pages with lots of contents, rather hard.

It's a bug and it will be solved in https://github.com/rust-lang/rust/pull/115682.

Please open an issue and link to this PR. I think the concern is valid, and it might be possible to add a mechanism to allow to disable this behaviour, but it needs to be discussed first about:
 * Do we want to allow people to opt out of this feature?
 * How do we disable this feature?
   * A global attribute?
   * A command line option?
   * An attribute to be used on each type alias?
   * An attribute to be used on module?

Of course, any of the item above can be combined with others. Lot of possibilities. ;)

**Pull request merged**